### DISCUSSION TITLE
Parsing val.py runs to clearml

### DISCUSSION THREAD

**Discussion by Sodek93** [ultralytics/yolov5#9782] [Q&A]
Hi, 
I am training custom data with train.py and parse the results to a local clearml client. Works Fine!

Now i need to also push the results of the val.py runs to clearml. 
I use Clearml to have a overwiew of my Metrics to compare different Settings/dataset.yaml's.

When i call 
`python val.py --img 640 --data dataset.yaml --weights taskrunbest.pt --task test`
the validation runs but ClearMl doesn't catch any information of my validation run.

In addition i am missing the "results" file in the validation folder. I receive F1_Curve,P_Curve etc. but no mAP at all.

**Comment by glenn-jocher** [comment]
@Sodek93 hi thanks for the question! Experiment logging generally only applies to training models, not to running inference or validation, so you should not expect val.py to log any results to ClearML.

@thepycoder any thoughts here?

**Reply by thepycoder** [reply]
Several things here,

The fact that @Sodek93 doesn't receive mAP at all is not ClearML related? I'm not sure.

It's also quite reasonable to want to save validation output if you want to compare multiple dataset yamls etc.
I see 2 options here:

1. We build a similar interface to `train.py` with e.g. 1 callback at the end that allows a logger to report the final results, plots, metrics etc. This could just be a callback function that's added in the `__init__.py` which calls the respective loggers util files. I'm assuming a single report at the end would be enough, as I don't see the reason to have e.g. iteration logging.
2. @Sodek93 and others that want this can add the respective logger statements into `val.py` themselves, but we could make it easier with an entry in the readme or something. In the end, `val.py` is much less complex than `training.py` and adding the necessary code directly will probably be relatively simple.

Personally, I don't see anything wrong with option 1 per se, the loggers `__init__` would initialize the loggers that are installed and then 1 function call would wrap everything up. Am I thinking too simple here? It seems to boil down to how much work you're willing to put into it :)

**Reply by glenn-jocher** [reply]
@thepycoder got it. I've actually not come across a user asking for this feature before, so was wondering if you guys had seen similar workflows before. I think let's wait for now and see if we get any other similar feature requests.

**Reply by thepycoder** [reply]
That's fair :) @Sodek93 do you think you can add custom logs to the `val.py` script yourself? It shouldn't be much more than adding a `task = Task.init(...)` at the top and maybe a `task.get_logger().report_scalar(...)` down in the metric section. The plots and parameters should be captured automatically :)

**Reply by Sodek93** [reply]
Hey,Thank you guys for your help!

I am trying to find the same code 
` task = Task.init(...)`
and
`task.get_logger().report_scalar(...)`
in train.py but can't find it.

i found this in clearml_utils.py:

`task.get_logger().report_scalar( title, series, value, iteration)`
But i am not sure what values the arguments should have... 

This would be the Task i create at the top in val.py
```
         clearmltask=Task.init(project_name='YOLOv5',
                   task_name='Validation',
                   tags=['YOLOv5'],
                   output_uri=True,
                   auto_connect_frameworks={'pytorch': False}
                   # We disconnect pytorch auto-detection, because we added manual model save points in the code)
                   )
```
I am a new user of Clearml and YoloV5 so i think i would need to spend a lot of more time in the code to get this right :S

**Reply by thepycoder** [reply]
Hey @Sodek93 
Sorry for the late reply.

How is the integration going?
The task init is indeed exactly what should be on top of the script, you were right about that!

For what concerns `task.get_logger().report_scalar( title, series, value, iteration)`
`title` is the title of the plot it will end up in. 1 `title` corresponds to 1 plot in the ClearML scalars tab.
`series` is the name of a trace in that plot. If you log 2 scalars, each with the same `title` but a different `series` name, you'll end up with a single plot with 2 traces on it in the scalars tab
`value` is the actual value, the point on the graph you want to capture, e.g. mAP. This will be the y-axis value
`iteration` is the epoch/step nr depending on how you want to track it. It will become the x-axis in the plot

Basically you give some numbers over time to ClearML and it will generate the plot for you in the interface based on these metrics.
